Mathlib/Topology/Instances/CantorSet.lean

```diff
@@ -0,0 +1,99 @@
+import Mathlib.Topology.Instances.Real
+
+/-!
+# Ternary Cantor Set
+
+This file defines the Cantor ternary set and proves a few properties.
+
+## Main Definitions
+
+* `preCantorSet n`: The order `n` pre-Cantor set, defined inductively.
+* `cantorSet`: The ternary Cantor set, defined as the intersection of all
+  pre-Cantor sets.
+-/
+
+/-- We define the preCantorSet as the preimages under the iterations of some functions. -/
+def preCantorSet : ℕ → Set ℝ
+  | 0 => Set.Icc 0 1
+  | n + 1 => (· / 3) '' preCantorSet n ∪ (fun x ↦ (2 + x) / 3) '' preCantorSet n
+
+@[simp] lemma preCantorSet_zero : preCantorSet 0 = Set.Icc 0 1 := rfl
+@[simp] lemma preCantorSet_succ (n : ℕ) :
+    preCantorSet (n + 1) = (· / 3) '' preCantorSet n ∪ (fun x ↦ (2 + x) / 3) '' preCantorSet n :=
+  rfl
+
+/-- The Cantor set is the intersection of all pre-Cantor sets. -/
+def cantorSet : Set ℝ := ⋂ n, preCantorSet n
+
+
+/-!
+## Simple Properties
+-/
+
+lemma quarters_mem_preCantorSet (n : ℕ) : 1/4 ∈ preCantorSet n ∧ 3/4 ∈ preCantorSet n := by
+  induction n with
+  | zero =>
+    simp only [preCantorSet_zero, inv_nonneg]
+    refine ⟨⟨ ?_, ?_⟩, ?_, ?_⟩ <;> linarith
+
+  | succ n ih =>
+    apply And.intro
+    · -- goal: 1/4 ∈ preCantorSet n
+      -- follows by the inductive hyphothesis, since 3/4 ∈ preCantorSet (n-1)
+      exact Or.inl ⟨3/4, ih.2, by simp only; linarith ⟩
+
+    · -- goal: 3/4 ∈ preCantorSet n
+      -- follows by the inductive hyphothesis, since 1/4 ∈ preCantorSet (n-1)
+      exact Or.inr ⟨1/4, ih.1, by simp only; linarith ⟩
+
+lemma quarter_mem_preCantorSet (n : ℕ) : 1/4 ∈ preCantorSet n := (quarters_mem_preCantorSet n).1
+
+theorem quarter_mem_cantorSet : 1/4 ∈ cantorSet := by
+  simp only [cantorSet,iInf, Set.sInf_eq_sInter, Set.sInter_range, Set.mem_iInter]
+  exact quarter_mem_preCantorSet
+
+lemma zero_mem_preCantorSet (n : ℕ) : 0 ∈ preCantorSet n := by
+  induction n with
+  | zero =>
+    simp [preCantorSet]
+  | succ n ih =>
+    exact Or.inl ⟨0, ih, by simp only [zero_div]⟩
+
+theorem zero_mem_cantorSet : 0 ∈ cantorSet := by simp [cantorSet, zero_mem_preCantorSet]
+
+/-- The ternary Cantor set is a subset of [0,1]. -/
+lemma cantorSet_subset_unitInterval : cantorSet ⊆ Set.Icc 0 1 := by
+  intro x hx
+  simp only [cantorSet, Set.iInf_eq_iInter, Set.mem_iInter] at hx
+  exact hx 0
+
+/-- The ternary Cantor set is closed. -/
+lemma isClosed_cantorSet : IsClosed cantorSet := by
+  let f := Homeomorph.mulLeft₀ (1/3:ℝ) (by norm_num)
+  let g := (Homeomorph.addLeft (2:ℝ)).trans f
+  apply isClosed_iInter
+  intro n
+  induction n with
+  | zero =>
+    exact isClosed_Icc
+  | succ n ih =>
+    refine IsClosed.union ?_ ?_
+    · refine (ClosedEmbedding.closed_iff_image_closed ?succ.refine_1.hf).mp ih
+      convert f.closedEmbedding using 2
+      simp [f, div_eq_inv_mul]
+    · refine (ClosedEmbedding.closed_iff_image_closed ?succ.refine_2.hf).mp ih
+      convert g.closedEmbedding using 2
+      simp [g, f, div_eq_inv_mul]
```

Comment:
it's generally undesirable to name goals with inscrutable names like `?succ.refine_1.hf`, and in this case, there's a simpler approach. Note that you can use dot notation to go directly from the homeomorphisms to closed embeddings.
```suggestion
    · simpa [f, div_eq_inv_mul] using f.closedEmbedding.closed_iff_image_closed.mp ih
    · simpa [g, f, div_eq_inv_mul] using g.closedEmbedding.closed_iff_image_closed.mp ih
```